### ISSUE TITLE
macos release seems broken

### ISSUE THREAD

**Issue by 2mol** [haskell/haskell-ide-engine#992]
Hi folks!

I was trying to troubleshoot HIE for somebody who reported issues getting it to work (for vscode) at https://github.com/alanz/vscode-hie-server/issues/126

Downloading the compiled release (yay for having that option now btw) will not work, seemingly some places have @bubba s path compiled in (`/Users/luke/Source/haskell-ide-engine/.stack-work/install/x86_64-osx/...`). See attached [4-Haskell HIE (pboy).log](https://github.com/haskell/haskell-ide-engine/files/2675888/4-Haskell.HIE.pboy.log)

I ran into some other smaller issues as well, I think polishing those up could make the installation experience accessible to a lot more people. I volunteer to work on documentation if we can agree on what to improve.

- ~~it wasn't obvious to me how I can peek into the logs between hie and vscode. What's the recommended way?~~ Found the logs nicely accessible in a dropdown under the output pane of vscode. Maybe we could put this in the Readme under the troubleshooting section at the bottom?
- it seems the `hie` executable is used instead of `hie-wrapper`, despite my impression that the latter would be used preferentially. On that note: Why have the `hie` executable there in the first place?
- I saw a complaint in the logs about `hlint.yaml` not being found, it's unclear if that was a warning or an error.

**Comment by Rhywun** [comment]
I tried to install v0.4.0.1 over v0.2.2.0 using the instructions under "Installation with stack" and something broke on this recommended instruction (after using brew to upgrade stack):

`stack install cabal-install`

Got "ExitFailure 1" trying to install GHC 8.6.2 but no indication of where to look for a log to explain it, so I gave up. Also, I broke HIE on my machine. I may try to start from scratch but will follow along here to see if anything develops.

**Comment by 2mol** [comment]
Hi @Rhywun! Let's split this into a separate issue since the cause will be different from the one affecting the precompiled binary release. I'll then try to redo my installation step by step and see if I can help your case!

**Comment by Rhywun** [comment]
@2mol 

Done, thanks

**Comment by jasonmcole** [comment]
The specific error message I got from make was:

```
dyld: Library not loaded: /usr/local/opt/gmp/lib/libgmp.10.dylib
  Referenced from: /Users/jasoncole/.stack/programs/x86_64-osx/ghc-8.6.2.temp/ghc-8.6.2/libraries/base/dist-install/build/libHSbase-4.12.0.0-ghc8.6.2.dylib
  Reason: image not found
make[1]: *** [install_packages] Abort trap: 6
make: *** [install] Error 2
```

Installing gmp with `brew install gmp` , then re-running `stack ./install.hs build-all` proceeded without issue.

**Comment by Anrock** [comment]
@jasonmcole that's known issue with ghc on osx. See [README](https://github.com/haskell/haskell-ide-engine#macos-got-error-while-installing-ghc-861-or-862---dyld-library-not-loaded-usrlocaloptgmpliblibgmp10dylib) and [ghc bug](https://gitlab.haskell.org/ghc/ghc/issues/15769)

**Comment by dboardmanrally** [comment]
Running the 8.6.4 install script attempts to install ghc 8.6.2 on mac OSX

```
$ stack ./install.hs hie-8.6.4
stack ./install.hs build-doc-8.6.4

Installing GHC ...
Received ExitFailure 2 when running
Raw command: /usr/bin/make install
Run from: /Users/drew.boardman/.stack/programs/x86_64-osx/ghc-8.6.2.temp/ghc-8.6.2/

Installing GHC ...
Received ExitFailure 2 when running
Raw command: /usr/bin/make install
Run from: /Users/drew.boardman/.stack/programs/x86_64-osx/ghc-8.6.2.temp/ghc-8.6.2/
```

**Comment by fendor** [comment]
Yes, the install script uses ghc 8.6.2. 
If this causes troubles for you, you can change the resolver in `shake.yaml` and in `install.hs`.
That problem, that you have to update two places for this, is currently being worked on and will be resolved soon.

**Comment by mouse07410** [comment]
@fendor could you please tell where in `install.hs` is the dependency on `ghc-8.6.2`? I could not find it.

**Comment by mouse07410** [comment]
Here are some logs from failed runs. Help would be greatly appreciated.
[cabal-build-all.txt](https://github.com/haskell/haskell-ide-engine/files/3110446/cabal-build-all.txt)
[cabal-hie-8.6.4.txt](https://github.com/haskell/haskell-ide-engine/files/3110447/cabal-hie-8.6.4.txt)
[cabal-hie-8.6.4a.txt](https://github.com/haskell/haskell-ide-engine/files/3110449/cabal-hie-8.6.4a.txt)
[cabal-hie-8.6.4b.txt](https://github.com/haskell/haskell-ide-engine/files/3110450/cabal-hie-8.6.4b.txt)

```
$ ghcup debug-info
Script version: 0.0.7

Script variables:
  GHC install location:     /Users/ur20980/.ghcup/ghc
  Binary install location:  /Users/ur20980/.ghcup/bin
  Tarball cache location:   /Users/ur20980/.ghcup/cache
  Downloader:               curl --fail -O <url>
  Script update url:        https://raw.githubusercontent.com/haskell/ghcup/master//ghcup
  GHC download baseurl:     https://downloads.haskell.org/~ghc
  Meta download url         https://raw.githubusercontent.com/haskell/ghcup/master//.download-urls
  Meta download format      1
  Meta version url          https://raw.githubusercontent.com/haskell/ghcup/master//.available-versions
  Meta version format       1

Detected system information:
  Architecture:   x86_64
  Distribution:   Darwin
  Distro alias:   darwin
  Distro version: 18.5.0
$ ghc -V
The Glorious Glasgow Haskell Compilation System, version 8.6.4
$ cabal --version
cabal-install version 2.4.1.0
compiled using version 2.4.1.0 of the Cabal library 
$ stack --version
Version 1.9.3, Git revision 40cf7b37526b86d1676da82167ea8758a854953b (6211 commits) x86_64 hpack-0.31.1
$ haddock --version
Haddock version 2.22.0, (c) Simon Marlow 2006
Ported to use the GHC API by David Waern 2006-2008
$ 
```

**Comment by fendor** [comment]
The `install.hs` depends implicitly on ghc-8.6.2 via its resolver in `shake.yaml`. You can safely bump it to a newer snapshot.

**Comment by mouse07410** [comment]
@fendor thanks. Edited `shake.yaml`, and got the build succeed on one machine (so far).

Could you explain the purpose of `stack-8.6.x.yaml` files vs. `stack.yaml`? Are they actually used? If so, how? I mean - do I need to adjust resolver in those that are version-relevant to my installation?

Here's config from the successful build - since two different resolvers seem specified, can you please tell me which one was actually used?
```
$ head stack-8.6.4.yaml 
resolver: lts-13.15 # GHC 8.6.4
packages:
- .
 . . . . .
```

```
$ head stack.yaml
resolver: nightly-2019-04-05 # GHC 8.6.4
packages:
- .
 . . . . .
```

```
$ head shake.yaml 
# Used to provide a different environment for the shake build script
#resolver: nightly-2018-12-15 # GHC 8.6.2
resolver: lts-13.15 # GHC 8.6.4
packages:
- .
 . . . . .
```

```
$ type hie
hie is hashed (/Users/ur20980/.local/bin/hie)
$ hie --version
Version 0.8.0.0 (2579 commits) x86_64 ghc-8.6.4
$ 
```

**Comment by power-fungus** [comment]
The different stack.yaml files are used to build `hie` with different `ghc` versions. `stack.yaml` is used to run `install.hs` in an isolated snapshot. 

At what version does the cabal build still fail? You could run the command manually with `cabal new-install exe:hie -w <ghc-path>` and check if any error is reported. (ghc-path used by `install.hs` can be found by calling `stack install.hs cabal-ghcs`)

**Comment by mouse07410** [comment]
> The different `stack.yaml` files are used to build hie with different `ghc` versions...

So, if (for example) I use only `ghc-8.6.3` or `ghc-8.6.4`, how should I run the build?

> At what version does the cabal build still fail?

Please see the complete logs I posted a couple of messages up - they're from yesterday's failed builds. Cabal build failed with the current (latest stable) Cabal, stack, and GHC-8.6.4 (and I think - 8.6.3 too). I haven't tried running Cabal manually (yet).

```
$ stack ./install.hs cabal-ghcs

******************************************************************
Found the following GHC paths: 
ghc-8.6.4: /Users/ur20980/.ghcup/bin/ghc-8.6.4
ghc-8.6.3: /Users/ur20980/.ghcup/bin/ghc-8.6.3

******************************************************************

Build completed in 0.00s
```

P.S. As of now, I have successful builds of the current master on three machines, two with GHC-8.6.4, and one with GHC-8.6.3. All - via stack only (no cabal yet).

**Comment by mouse07410** [comment]
Also, I want to thank the developers/maintainers of this repo, whose hard work made the build (at least in one form) succeed now, when the previous versions failed.

**Comment by power-fungus** [comment]
I have never seen your errors before. Please ensure a fresh repository:

```bash
git clean -Xdf && git pull # or do a fresh clone
git submodule sync && git submodule update --init
```

> So, if (for example) I use only ghc-8.6.3 or ghc-8.6.4, how should I run the build?

The officially supported method to install these two versions only is `stack install.hs hie-8.6.4 && stack install.hs hie.8.6.3 && stack install.hs build-doc`.

if these fail with a similar error message I am interested in the output of `stack build --stack-yaml=stack-8.6.4.yaml` and `cabal new-install --symlink-bindir=$HOME/.local/bin`, which are the main commands to build `hie`, which the `install.hs` wraps.

**Comment by mouse07410** [comment]
> I have never seen your errors before. Please ensure a fresh repository...

Thanks. Will try later, and report.

> The officially supported method to install [for 8.6.4] ... is `stack install.hs hie-8.6.4` ...

OK, thanks. Got it. In fact, that's how my installation succeeded.

> `stack install.hs build-doc`

This doesn't work:
```
$ hie --version
Version 0.7.0.0 (2483 commits) x86_64 ghc-8.6.4
$ stack ./install.hs build-doc
Error when running Shake build system:
  at want, called at src/Development/Shake/Internal/Args.hs:83:69 in shake-0.17.6-CIDPVLgp7pwBLb5kEJCmoZ:Development.Shake.Internal.Args
* Depends on: build-doc
  at error, called at src/Development/Shake/Internal/Rules/File.hs:180:58 in shake-0.17.6-CIDPVLgp7pwBLb5kEJCmoZ:Development.Shake.Internal.Rules.File
* Raised the exception:
Error, file does not exist and no rule available:
  build-doc


$ stack ./install.hs build-docs
# stack (for build-doc-8.2.1)
Downloaded nightly-2017-11-24 build plan.    
Compiler version mismatched, found ghc-8.6.4 (x86_64), but expected minor version match with ghc-8.2.1 (x86_64) (based on resolver setting in /Users/ur20980/src/haskell-ide-engine/stack-8.2.1.yaml).
To install the correct GHC into /Users/ur20980/.stack/programs/x86_64-osx/, try running "stack setup" or use the "--install-ghc" flag.
Error when running Shake build system:
  at want, called at src/Development/Shake/Internal/Args.hs:83:69 in shake-0.17.6-CIDPVLgp7pwBLb5kEJCmoZ:Development.Shake.Internal.Args
* Depends on: build-docs
  at need, called at install.hs:80:25 in main:Main
* Depends on: build-doc-8.2.1
* Raised the exception:
user error (Development.Shake.command_, system command failed
Command: stack --stack-yaml=stack-8.2.1.yaml install hoogle
Exit code: 1
Stderr:
Downloaded nightly-2017-11-24 build plan.    
Compiler version mismatched, found ghc-8.6.4 (x86_64), but expected minor version match with ghc-8.2.1 (x86_64) (based on resolver setting in /Users/ur20980/src/haskell-ide-engine/stack-8.2.1.yaml).
To install the correct GHC into /Users/ur20980/.stack/programs/x86_64-osx/, try running "stack setup" or use the "--install-ghc" flag.
)


$ stack ./install.hs build-doc-8.6.4
# stack (for build-doc-8.6.4)
Copying from /Users/ur20980/src/haskell-ide-engine/.stack-work/install/x86_64-osx/lts-13.15/8.6.4/bin/hoogle to /Users/ur20980/.local/bin/hoogle

Copied executables to /Users/ur20980/.local/bin:
- hoogle
# stack (for build-doc-8.6.4)
Starting generate
Reading Cabal... 7.58s
[156/2343] data-ordlist... 0.01s                    
[284/2343] network-house... 0.05s             
[414/2343] regex-applicative... 0.03s 
. . . . .
[2232/2343] hledger-web... 0.53s        
[2271/2343] hoogle... 0.00s                
Packages missing documentation: abstract-deque-tests al array-memoize base-noprelude bin-package-db bno055-haskell bytestring-builder doctest-discover either-both ghc-heap h2c haskell-gi-overloading hfsevents inliterate interpolator io-manager ipython-kernel lame lens-simple linux-file-extents modular pattern-arrows proto-lens-protobuf-types rocksdb-haskell rocksdb-query rts secp256k1-haskell telegram-bot-simple tensors ucam-webauth ucam-webauth-types Win32-notify zeromq4-patterns
Found 48063 warnings when processing items

Reordering items... 0.28s
Writing tags... 2.29s
Writing names... 1.64s
Writing types... 9.38s
Took 1m45s
Build completed in 1m48s

$
```
Any idea why the successful doc build above complains about missing documentation?

> `stack build --stack-yaml=stack-8.6.4.yaml`
Also, to clarify the use of `stack-8.x.x.yaml`: does `install.hs` actually use them when a specific GHC version is requested? What's the relationship between `stack.yaml` and `stack-8.6.4.yaml`?

**Comment by power-fungus** [comment]
You seem your repository could benefit from `git pull`

**Comment by mouse07410** [comment]
You may not believe me, but I'm doing "git pull" and "git submodule update --init" every time I start doing anything on this repo. Good hygiene, y'know...

**Comment by fendor** [comment]
`hie --version` hints to an old version of hie, since the current version is `0.8.0.0`. Or there is an old hie version on the PATH.
The target `build-docs` does not exist on HEAD anymore.
Thus, your repository is either not up-to-date or you are not following the master branch.

To the relation, they are not related. `stack.yaml` is just using the nightly stackage snapshot. It is used for development and CI. The install script does afaik not use it at all.

**Comment by mouse07410** [comment]
Thank you - I just noticed that on one machine indeed it built (or used) an old version (0.7 instead of 0.8). I'll take care of that. 

So, it I'm using "stack install.hs ..." and want to ensure the build users a specific resolver and GHC-8.6.4, I need to edit shake.yaml and stack-8.6.4.yaml?

**Comment by fendor** [comment]
Afaik you only have to edit the `shake.yaml` such that install.hs is run with the ghc version you wish. If you use `stack install.hs hie-8.6.4` the correct `stack*.yaml` will be automatically used for the `hie` executable.

**Comment by mouse07410** [comment]
Great! But do I need to make sure stack-8.6.4.yaml doesn't pull a different resolver?

**Comment by fendor** [comment]
No, the only additional resolver is pulled in by `shake.yaml`. Everything else should work as expected.

**Comment by mouse07410** [comment]
> ...or you are not following the master branch...

Indeed. That machine was on a different branch. Thanks for pointing that out. Just fixed.
```
$ hie --version
Version 0.8.0.0 (2589 commits) x86_64 ghc-8.6.4
$
```

**Comment by jneira** [comment]
This seems to be fixed, closing
